File: GUI/WpfMidiOrganPedals/Devices/ConfigurationRequestMessage.cs
using System.Linq;

namespace WpfMidiOrganPedals.Devices
{
    internal class ConfigurationRequestMessage : ConfigurationMessageBase
    {
        public const byte Id = 0x03;

        public ConfigurationRequestMessage(int firstNote, int velocity, int debouncingTime, int[] pedalPins)
        {
            this.firstNote = (byte)firstNote;
            this.velocity = (byte)velocity;
            this.debouncingTime = (byte)debouncingTime;
            this.pedalPins = pedalPins.Select(x => (byte)x).ToArray();
        }

        public override RawMessage Pack()
        {
            return Pack(Id);
        }
    }
}

File: Firmware/ConfigurationRequestMessage.cpp
#include "ConfigurationRequestMessage.h"
#include "RawMessageExtractor.h"
#include "Misc.h"


bool ConfigurationRequestMessage::Unpack(const RawMessage &input)
{
  RawMessageExtractor extractor(input.GetData(), input.GetSize());
  
  if (!extractor.Get(FirstNote)) {
    return false;
  }
  
  if (!extractor.Get(Velocity)) {
    return false;
  }
  
  if (!extractor.Get(DebouncingTime)) {
    return false;
  }
  
  if (!extractor.Get(PedalPins, ARRAY_SIZE(PedalPins))) {
    return false;
  }
  
  if (!extractor.Finish()) {
    return false;
  }
  
  return true;
}

